### PR TITLE
Implement String#split with block

### PR DESCRIPTION
**Pull request by splattael**
Hi,

this PR enhances all flavors of `String#split` with a block.

After this PR, you can pass a block to `String#split`:

```crystal
"foo bar".split { |chunk| p chunk }
"foo bar".split(' ') { |chunk| p chunk }
"foo bar".split(" ") { |chunk| p chunk }
"foo bar".split(/ /) { |chunk| p chunk }
```

Additional tests were not added because all non-block versions of `String#split` exercise `String#split(..., &block)`. (Only one test for an edge case was added, though)

Kind regards,
  Peter

### CODE REVIEW

**Comment by kostya** [comment]
https://github.com/crystal-lang/crystal/pull/2541

**Comment by splattael** [comment]
@kostya Dammit, should have searched for it before ;)

Sorry, for the noise.

(I am actually using the block version to speed-up things...)

**Comment by kostya** [comment]
it was not merged, and i also want a version which yield slice, for speed

**Comment by splattael** [comment]
@kostya Yeah, my version yields `String`s which is slower than `Slice` but still faster than `"foo bar".split(' ').each { ... }`.

Anyhow, if it's not wanted, I'll keep my monkey patch^W^Wextension for now ;)

File: spec/std/string_spec.cr [hunk 1]
```diff
@@ -738,6 +738,7 @@ describe "String" do
       assert { "   foo  bar".split(' ').should eq(["", "", "", "foo", "", "bar"]) }
       assert { "   foo   bar\n\t  baz   ".split(' ').should eq(["", "", "", "foo", "", "", "bar\n\t", "", "baz", "", "", ""]) }
       assert { "   foo   bar\n\t  baz   ".split.should eq(["foo", "bar", "baz"]) }
+      assert { "   foo   bar\n\t  baz   ".split(1).should eq(["   foo   bar\n\t  baz   "]) }
```
Comment (splattael) [inline spec/std/string_spec.cr]: @asterite Even if #3702 won't be merged maybe you could :cherries: this test?

**Comment by ysbaddaden** [comment]
I'm reopening the PR. The problem of #2541 is avoided because it yields String instead of an unexpected Slice —which would fit better as `Slice#split` and be just as awesome. Maybe it could implement `Slice#split(Char)` and use it, but that wouldn't cover other cases (String, Regex), and I think this is a great improvement already (that could be further improved).

String :+1: from me!

**Comment by kostya** [comment]
yielding slice is ok, because it internal method which i need to use for speed, common method still return strings.

**Comment by asterite** [comment]
I'm also OK with this addition.

Eventually I'd also like to be able to get an `Iterator` from a split. Maybe it can be named `each_split`.

@kostya We could add a separate method that yields slices, with a different name (maybe `each_split_slice` or something like that). Slices are much more low-level and needed when you want to squeeze performance in case of a bottleneck, and Crystal should allow you do do that, but the default is usually "user-friendly" (String is better than Slice, Array is better than Slice and StaticArray, etc.)

File: src/string.cr [hunk 2]
```diff
@@ -2441,11 +2441,35 @@ class String
...
+  #
+  # ```
+  # ary = [] of String
+  # old_pond = "
+  #   Old pond
+  #   a frog leaps in
+  #   water's sound
+  # "
+  # old_pond.split { |s| ary << s }; ary    # => ["Old", "pond", "a", "frog", "leaps", "in", "water's", "sound"]
+  # old_pond.split(3) { |s| ary << s }; ary # => ["Old", "pond", "a frog leaps in\n  water's sound\n"]
```
Comment (asterite) [inline src/string.cr]: `ary` should probably cleared before each split. This applies to this example and other examples below.

**Comment by splattael** [comment]
@asterite Ready to merge. Do you want me to squash the commits?

**Comment by asterite** [comment]
@splattael No problem, I think the separate commits look good :-)

I'll merge it soon. Thank you!
